frontend/catalyst/jax_extras/tracing.py

```diff
@@ -248,6 +312,97 @@ def __init__(self, boxid: int, e: JaxprEqn):
+        """Return the number of implicit input arguments of the function"""
+        return len([() for _, k in self.in_type if not k])
+
+
+@dataclass
+class OutputSignature:
+    """Meta-parameters of a function which become available after the tracing to the function is
+    complete."""
+
+    out_jaxpr: Callable[[], ClosedJaxpr]
+    out_type: Callable[[], OutputType]
+    out_consts: Callable[[], list]
+    out_tree: Callable[[], PyTreeDef]
+    out_initial_jaxpr: Callable[[], Jaxpr]
+
+    def num_implicit_outputs(self) -> int:
+        """Return the number of implicit resuts of the function"""
+        return len([() for _, k in self.out_type() if not k])
+
+
+def deduce_signatures(
+    f: Callable, args, kwargs, expansion_strategy
+) -> Tuple[Callable, InputSignature, OutputSignature]:
+    """Prepares the callable ``f`` for tracing by wrapping it into a WrappedFun container accepting
+    expanded flattened arguments and returning expanded flatten results. Jax input and output types
+    are returned along with the other related information aggregated into input and output signature
+    datatypes.
+
+    Args:
+        f (Callable): Python function to trace. By definition, the function accepts collapsed
+                      unflattened arguments and returns collapsed unflattened results.
+        args (Any): Positional arguments to be passed to ``f``, typically Jax tracers.
+        kwargs (dict): Keyword arguments to be passed to ``f``.
+        expansion_strategy (ExpansionStrategy): Argument and result expansion settings to use when
+                                                expanding arguments and results.
+
+    Returns:
+        Callable: Python function, accepting flattened expanded arguments and returning flattened
+                  expanded results
+        InputSignature: Input signature of the function
+        OutputSignature: Output signature of the function. Fields of the output signature are
+                         available after exiting from the Callable.
+    """
+    flat_args, in_tree = tree_flatten((args, kwargs))
+    trace: DynamicJaxprTrace = find_top_trace(flat_args)
+    flat_tracers = [trace.full_raise(a) for a in flat_args]
+    # flat_axes_specs = _flat_axes_specs(abstracted_axes, *args, **kwargs)
+    in_expanded_args, in_type = expand_args(
+        flat_tracers, expansion_strategy=expansion_strategy  # axes_specs=flat_axes_specs,
+    )
+    wf = wrap_init(f)
+    wf, out_tree_promise = flatten_fun(wf, in_tree)
+    wf, out_sig_promise = expanded_fun(wf, (in_type, expansion_strategy))
+    wf = annotate(wf, in_type)
+    return (
+        wf,
+        InputSignature(in_type, in_tree, in_expanded_args),
+        OutputSignature(
+            lambda: ClosedJaxpr(convert_constvars_jaxpr(out_sig_promise()[0]), ()),
+            lambda: out_sig_promise()[1],
+            lambda: out_sig_promise()[2],
+            out_tree_promise,
+            lambda: out_sig_promise()[0],
```

Comment:
Why do these need to be functions?

Reply:
It is another feature of this Jax transformation engine. They allow transformations to return aux values as thunks, which are functions. If I remember it correctly, they rise an exception if called before the actual wrapped function is called. Here we collect two aux results into a single structure so I decided to follow this style without much changes.

Reply:
Is it possible to document a bit more what each `out_sig_promise` output is by adding a variable name to them? Even if they are still functions. E.g., `get_new_dynamic_args = lambda: out_sig_promise()[0]` and then use it here? Without variable names I don't really have a good picture of what these functions do.

Reply:
OK, done. Also added an extended comment to the dataclasses.